src/lib/mina_base/transaction_logic.ml

```diff
@@ -1277,46 +1265,13 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
         Party.Predicated.t) (a : Account.t) : (Account.t, _) Result.t =
     let open Snapp_basic in
     let open Result.Let_syntax in
-    let%bind balance =
-      let%bind b = add_signed_amount a.balance balance_change in
-      let fee = constraint_constants.account_creation_fee in
-      let%bind () =
-        (* TODO: Fix when we want to enable tokens. The trickiness here is we need to subtract
-           the account creation fee from somewhere (like the fee excess in the local state) *)
-        if Token_id.(equal default) token_id then Ok ()
-        else Error Transaction_status.Failure.Cannot_pay_creation_fee_in_token
-      in
-      if is_new then
-        Balance.sub_amount b (Amount.of_fee fee)
-        |> opt_fail Amount_insufficient_to_create_account
-      else Ok b
-    in
-    (* Check send/receive permissions *)
-    let%bind () =
-      if Amount.(equal zero) balance_change.magnitude then Ok ()
-      else
-        check Update_not_permitted_balance
-          (check_auth
-             ( match balance_change.sgn with
-             | Pos ->
-                 a.permissions.receive
-             | Neg ->
-                 a.permissions.send ))
-    in
     (* Check timing. *)
     let%bind timing =
       match balance_change.sgn with
       | Pos when not is_new ->
           Ok a.timing
       | _ ->
-          let txn_amount =
-            match balance_change.sgn with
-            | Pos ->
-                Amount.zero
-            | Neg ->
-                balance_change.magnitude
-          in
-          validate_timing ~txn_amount
+          validate_timing ~txn_amount:Amount.zero
```

Comment:
shouldn't `txn_amount` be the magnitude in the case of balance deduction?

Reply:
We want to handle it as a distinct failure with specialised logic. We have to calculate the resulting balance again anyway, so we do it ahead of time in the `balance` calculation. This just matches what was already done on the snark side.